### PR TITLE
Integrate LibFuzzer

### PR DESCRIPTION
**Pull request by henrybear327**
By leveraging the LLVM's LibFuzzer, we will be able to test our emulator code more thoroughly.

The seed corpus being added automatically are all the elf files in the build folder.

Major changes:
- Conditional compilation for the main function, as the fuzzer will generate its own main function
- Add Clang to the Dockerfile 
- Introduce loading binary stream as elf file, as the input from fuzzer will be passed into the emulator directly instead of going through a file

Minor changes:
- Fixed ELF verification logic 

Currently, known improvements are:
- Fix the Makefile to handle the compilation flags and compile coexisting .c and .cc files correctly

### CODE REVIEW

**Comment by henrybear327** [comment]
Based on the initial fuzzer runs, I think the ELF loader logic doesn't check for invalid ELF strictly yet. 

For example, this is one of the crashes produced by the fuzzer
```
Executed with failure
src/elf.c:145:52: runtime error: member access within misaligned address 0x1000078803437 for type 'const struct Elf32_Shdr', which requires 4 byte alignment
0x1000078803437: note: pointer points here
<memory cannot be printed>
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/elf.c:145:52 in 
src/elf.c:145:52: runtime error: load of misaligned address 0x1000078803437 for type 'const Elf32_Word' (aka 'const unsigned int'), which requires 4 byte alignment
0x1000078803437: note: pointer points here
<memory cannot be printed>
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/elf.c:145:52 in 
AddressSanitizer:DEADLYSIGNAL
```

This [PR](https://github.com/sysprog21/rv32emu/pull/265) was also created based on the fuzzer's crash log.

File: Makefile [hunk 1]
```diff
@@ -3,6 +3,7 @@ include mk/toolchain.mk
 
 OUT ?= build
 BIN := $(OUT)/rv32emu
+FUZZER_BIN := $(OUT)/rv32emu_fuzzer
```
Comment (jserv) [inline Makefile]: The filename should come with `fuzz` suffix.

File: run_fuzzer.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+set -e -u -o pipefail
+
+echo "Compiling..."
+clang \
```
Comment (jserv) [inline run_fuzzer.sh]: Check the status of `clang`.

File: src/elf.c [hunk 3]
```diff
@@ -344,6 +359,11 @@ bool elf_open(elf_t *e, const char *path)
 #endif /* USE_MMAP */
 
     /* point to the header */
+    if (sizeof(struct Elf32_Ehdr) > e->raw_size) {
+        // this will never be a valid ELF file
```
Comment (jserv) [inline src/elf.c]: Use C-style comments. i.e., `/* comment */`.

File: src/fuzzer/fuzzer.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# Fuzzing
```
Comment (jserv) [inline src/fuzzer/fuzzer.md]: Move this file to `docs` directory.

File: src/riscv.h [hunk 5]
```diff
@@ -140,6 +140,14 @@ typedef struct {
     bool allow_misalign;
 } riscv_io_t;
 
+#ifdef FUZZER
+// For fuzzing only: initialize the RISC-V emulator, and execute the given
+// binary stream
+int rv_init_and_execute_binary(const uint8_t *data,
```
Comment (jserv) [inline src/riscv.h]: There is no need to prepare fuzzer specific API.

**Review by jserv — CHANGES_REQUESTED** [review]
Resolve the issues reported by Codacy.

File: src/riscv.h (hunk 5 above)
Comment (henrybear327) [reply src/riscv.h]: How can we open a binary stream with the current system set up though?

File: docs/fuzzer.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# Fuzzing
+
+We are using the [LLVM Fuzzer](https://llvm.org/docs/LibFuzzer.html).
+
+The fuzzer used here is without structured input generation. Instead, we rely
```
[the remark this answers is not in the document]
Comment (henrybear327) [reply docs/fuzzer.md]: Not sure what this is about.

File: src/riscv.h (hunk 5 above)
Comment (henrybear327) [reply src/riscv.h]: Maybe I will do the following:
- Take the entire logic out of the main function
- Send in the fuzzer input by using the `argv`

**Comment by henrybear327** [comment]
> Resolve the issues reported by Codacy.

I tried to resolve most, but there are some that are hard to understand what the error is about, or, some don't make sense to fix (single exit point per function, use the output of fprintf, don't use `abort()` in fuzzer, etc.) .

File: src/riscv.h (hunk 5 above)
Comment (jserv) [reply src/riscv.h]: > * Take the entire logic out of the main function
> * Send in the fuzzer input by using the `argv`

It sounds better. Again, `riscv.h` is considered as a public header.

File: .ci/riscv-fuzz.sh [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+	src/emulate.c \
+	src/riscv.c \
+	src/elf.c \
+	src/cache.c \
+	src/mpool.c \
+	src/main.c
+
+# populate the initial CORPUS for the fuzzer using valid elf
+mkdir -p build/fuzz/CORPUS_DIR
+cp build/*.elf build/fuzz/CORPUS_DIR
```
Comment (jserv) [inline .ci/riscv-fuzz.sh]: Don't copy all ELF files. Instead, you shall select some of them, which have been validated.

File: docs/fuzzer.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jserv) [reply docs/fuzzer.md]: Add some annotation to ignore?

File: docs/fuzzer.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+We are using the [LLVM Fuzzer](https://llvm.org/docs/LibFuzzer.html).
+
+The fuzzer used here is without structured input generation. Instead, we rely 
+on the fuzzer to mutate the input.
+
+The initial seeds are all the ELF files in the build folder.
+
+## Execution
+
+- `./.ci/riscv-fuzz.sh`
```
Comment (jserv) [inline docs/fuzzer.md]: Use shorter form. e.g., `.ci/fuzz.sh`

File: src/elf.c [hunk 9]
```diff
@@ -290,13 +290,29 @@ bool elf_load(elf_t *e, riscv_t *rv, memory_t *mem)
     return true;
 }
 
+#ifdef FUZZER
+bool elf_open(elf_t *e, const char *path, const uint8_t *data, size_t len)
```
Comment (jserv) [inline src/elf.c]: Why did you change the function prototype?

File: src/fuzz_target.cc [hunk 10]
```diff
@@ -0,0 +1,31 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include "riscv.h"
+
+const int max_instructions = 5000;
+
+// In order to be able to inspect a coredump we want to
+// crash on every ASAN error.
+extern "C" void __asan_on_error()
```
Comment (jserv) [inline src/fuzz_target.cc]: Why did you specify this file as C++ source instead of C?

Comment (henrybear327) [reply src/fuzz_target.cc]: I followed [the steps](https://llvm.org/docs/LibFuzzer.html) on the official site, and it uses C++ code. 

From the stacktrace on error, I believe that the LibFuzzer is implemented in C++. 

```
#0 0x55a9931f3b21 in __sanitizer_print_stack_trace (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0x19eb21) (BuildId: def03160f034cb06e8fb39819143fae47fa8a649)
    #1 0x55a9931663b8 in fuzzer::PrintStackTrace() (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0x1113b8) (BuildId: def03160f034cb06e8fb39819143fae47fa8a649)
    #2 0x55a99314bda9 in fuzzer::Fuzzer::AlarmCallback() (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0xf6da9) (BuildId: def03160f034cb06e8fb39819143fae47fa8a649)
    #3 0x7f9c66c[42](https://github.com/sysprog21/rv32emu/actions/runs/6905735139/job/18789172120#step:4:43)51f  (/lib/x86_64-linux-gnu/libc.so.6+0x4251f) (BuildId: a[43](https://github.com/sysprog21/rv32emu/actions/runs/6905735139/job/18789172120#step:4:44)bfc8428df6623cd498c9c0caeb91aec9be4f9)
    #4 0x55a99316499a in __sanitizer_cov_trace_const_cmp1 (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0x10f99a) (BuildId: def03160f034cb06e8fb39819143fae47fa8a649)
    #5 0x55a993273a78 in do_csw /home/runner/work/rv32emu/rv32emu/src/rv32_template.c:852:1
    #6 0x55a99323db7d in rv_step /home/runner/work/rv32emu/rv32emu/src/emulate.c:1029:13
    #7 0x55a9932c69a2 in run /home/runner/work/rv32emu/rv32emu/src/main.c:89:5
    #8 0x55a9932c69a2 in rv_init_and_execute_binary /home/runner/work/rv32emu/rv32emu/src/main.c:253:5
    #9 0x55a99322[47](https://github.com/sysprog21/rv32emu/actions/runs/6905735139/job/18789172120#step:4:48)df in fuzz_elf_loader(unsigned char const*, unsigned long) /home/runner/work/rv32emu/rv32emu/src/fuzz_target.cc:20:15
    #10 0x55a9932247df in LLVMFuzzerTestOneInput /home/runner/work/rv32emu/rv32emu/src/fuzz_target.cc:30:5
    #11 0x55a99314d3c3 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0xf83c3) (BuildId: def03160f034cb06e8fb39819143fae47fa8a6[49](https://github.com/sysprog21/rv32emu/actions/runs/6905735139/job/18789172120#step:4:50))
    #12 0x55a99314cb19 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0xf7b19) (BuildId: def03160f034cb06e8fb39819143fae47fa8a649)
    #13 0x55a99314e788 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile> >&) (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0xf9788) (BuildId: def03160f034cb06e8fb39819143fae47fa8a649)
    #14 0x55a99314ec72 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile> >&) (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0xf9c72) (BuildId: def03160f034cb06e8fb39819143fae47fa8a649)
    #15 0x55a99313cfc2 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0xe7fc2) (BuildId: def03160f034cb06e8fb39819143fae47fa8a649)
    #16 0x[55](https://github.com/sysprog21/rv32emu/actions/runs/6905735139/job/18789172120#step:4:56)a993166cb2 in main (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0x111cb2) (BuildId: def03160f034cb06e8fb39819143fae47fa8a649)
    #17 0x7f9c66c29d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: a43bfc8428df6623cd498c9c0caeb91aec9be4f9)
    #18 0x7f9c66c29e3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: a43bfc8428df6623cd498c9c0caeb91aec9be4f9)
    #19 0x55a993131a04 in _start (/home/runner/work/rv32emu/rv32emu/build/rv32emu_fuzz+0xdca04) (BuildId: def03160f034cb06e8fb39819143fae47fa8a649)
```

File: src/elf.c (hunk 9 above)
Comment (henrybear327) [reply src/elf.c]: The main purpose was to input the binary string from the fuzzer directly instead of having to go through a temporary file. 

I had an implementation with `fmemopen`, which avoided the temporary file.

File: docs/fuzzer.md (hunk 8 above)
Comment (henrybear327) [reply docs/fuzzer.md]: Updated!

I was following the convention in the `.ci` folder, thus added the `riscv-` prefix.

File: .ci/riscv-fuzz.sh (hunk 7 above)
Comment (henrybear327) [reply .ci/riscv-fuzz.sh]: May I ask what you mean by "validated"?

File: docs/fuzzer.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (henrybear327) [reply docs/fuzzer.md]: OK, I figured it out. It's the trailing white space when I did the linebreak to make the line <= 80 chars.

File: .ci/fuzz.sh [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    -include src/common.h \
+    -D RV32_FEATURE_EXT_F=0 \
+    -D RV32_FEATURE_SDL=0 \
+    -D DEFAULT_STACK_ADDR=0xFFFFE000 \
+    -D DEFAULT_ARGS_ADDR=0xFFFFF000 \
+    -D FUZZER \
+    -o build/rv32emu_fuzz \
+    src/fuzz_target.cc \
+    src/map.c \
+	src/utils.c \
```
Comment (jserv) [inline .ci/fuzz.sh]: Indent.

File: Makefile [hunk 12]
```diff
@@ -3,6 +3,7 @@ include mk/toolchain.mk
 
 OUT ?= build
 BIN := $(OUT)/rv32emu
+FUZZER_BIN := $(OUT)/rv32emu_fuzz
```
Comment (jserv) [inline Makefile]: should be `FUZZ_BIN`.

File: docs/fuzzer.md [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+We are using the [LLVM Fuzzer](https://llvm.org/docs/LibFuzzer.html).
+
+The fuzzer used here is without structured input generation. Instead, we rely
+on the fuzzer to mutate the input.
+
+The initial seeds are all the ELF files in the build folder.
+
+## Execution
+
+- `./.ci/fuzz.sh`
```
Comment (jserv) [inline docs/fuzzer.md]: No, simply refer to `.ci/fuzz.sh` without slash dot.

File: src/elf.c [hunk 14]
```diff
@@ -290,13 +290,29 @@ bool elf_load(elf_t *e, riscv_t *rv, memory_t *mem)
...
 bool elf_open(elf_t *e, const char *path)
+#endif
 {
     /* free previous memory */
     if (e->raw_data)
         release(e);
 
-#if defined(USE_MMAP)
+#if defined(FUZZER)
+    if ((data == NULL) || (len == 0)) {
```
Comment (jserv) [inline src/elf.c]: Rewrite as `if (!data || !len) {`

File: src/elf.c [hunk 15]
```diff
@@ -290,13 +290,29 @@ bool elf_load(elf_t *e, riscv_t *rv, memory_t *mem)
...
+        // if the fuzzer sent in an empty buffer, we don't proceed further
```
Comment (jserv) [inline src/elf.c]: Always use C-style comments.

File: .ci/fuzz.sh [hunk 16]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+set -e -u -o pipefail
+
+# compile
```
Comment (jserv) [inline .ci/fuzz.sh]: Refine this comment. Your intention was to check the availability of clang.

File: docs/fuzzer.md [hunk 17]
```diff
@@ -0,0 +1,21 @@
+# Fuzzing
+
+We are using the [LLVM Fuzzer](https://llvm.org/docs/LibFuzzer.html).
+
+The fuzzer used here is without structured input generation. Instead, we rely
+on the fuzzer to mutate the input.
+
+The initial seeds are all the ELF files in the build folder.
```
Comment (jserv) [inline docs/fuzzer.md]: Always use "directory" instead of "folder" in UNIX.

File: .ci/fuzz.sh [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+command -v clang &> /dev/null
+
+make clean
+clang \
```
Comment (jserv) [inline .ci/fuzz.sh:9]: It would be better to integrate fuzzing specific rules into the existing `Makefile`, so that we don't have to maintain.

**Comment by jserv** [comment]
Check https://docs.github.com/en/code-security/code-scanning/troubleshooting-sarif-uploads/results-exceed-limit and resolve the issues reported as "Code Scanning could not process the submitted SARIF file: rejecting SARIF"

File: .ci/fuzz.sh (hunk 18 above)
Comment (henrybear327) [reply .ci/fuzz.sh:9]: I tried in my initial prototyping, and it failed to compile properly and broke the existing makefile functionalities. 

I will revisit this point later.

File: docs/fuzzer.md [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
+The initial seeds are all the ELF files in the `build` directory.
+
+## Execution
+
+- `.ci/fuzz.sh`
+
+## References
+
+> Inspired by the fuzzer from [this project](https://github.com/fwsGonzo/libriscv/tree/master/fuzz).
```
Comment (jserv) [inline docs/fuzzer.md]: Explicitly mention "libriscv" in the text.

File: .ci/fuzz.sh (hunk 18 above)
Comment (jserv) [reply .ci/fuzz.sh:9]: Check 'sanitizer' target in jitboy: https://github.com/sysprog21/jitboy/blob/master/Makefile#L49-L51

**Comment by henrybear327** [comment]
> Check https://docs.github.com/en/code-security/code-scanning/troubleshooting-sarif-uploads/results-exceed-limit and resolve the issues reported as "Code Scanning could not process the submitted SARIF file: rejecting SARIF"

I removed directories that we are not interested in being scanned against, and it works for now. (I am not 100% sure what the root cause is, but the current fix seems to be sufficient enough already)

**Comment by henrybear327** [comment]
> Check https://docs.github.com/en/code-security/code-scanning/troubleshooting-sarif-uploads/results-exceed-limit and resolve the issues reported as "Code Scanning could not process the submitted SARIF file: rejecting SARIF"

After spending hours looking into the issues, I am still not able to resolve the problem properly, as the link above and the relative docs/issues that I tried to look up is all about `codeql` but not `codacy`. 

I have currently only patched the issue with a config file that excludes paths that we are not interested in code scanning.

File: src/elf.c (hunk 9 above)
Comment (henrybear327) [reply src/elf.c]: So I did a refactoring on the way that fuzzer's input is passed into the emulator.

- The entire logic is extracted from the main function into a new function called `rv_init_and_execute_elf`. The function prototype is added to `riscv.h`, as adding to `riscv-private.h` will cause the error "function symbol not found" during linking.
- `elf_open()` signature is still modified, as we aren't generating a tmp file to store the elf output, it's hard to reuse the current code (fuzzing will be a lot of IO if we write to disk, as the input will be mutated quickly)
-  The fuzzer will pass the input into the function `rv_init_and_execute_elf` using args. The input will then be parsed by the `parse_args`, but with a catch. As `getopt` can't deal with the input data that fuzzer generates (actually, even existing ELF files will cause the `getopt` to behave wrongly) since we have a lot of control sequence / non-ascii code points in the input. Thus, a special get opt loop is added. But the idea of sharing the same function interface by using `argc` and `args` is achieved.

File: .ci/fuzz.sh (hunk 18 above)
Comment (henrybear327) [reply .ci/fuzz.sh:9]: Will do it later! Thanks for the pointer.

**Comment by jserv** [comment]
> Flawfinder (reported by Codacy) found more than 10 potential problems in the proposed changes. Check the [Files changed](/sysprog21/rv32emu/pull/267/files) tab for more details.

@henrybear327, You should eliminate the warning "The atoi family of functions can potentially overflow or underflow integer values" reported by Codacy.

File: src/fuzz_target.cc [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+}
+extern "C" void __msan_on_error()
+{
+    abort();
+}
+
+static void fuzz_elf_loader(const uint8_t *data, size_t len)
+{   
+    int argc = 1 + 2 * 3 + 1;
+    char** args = (char**) malloc(sizeof(char*) * argc);
```
Comment (jserv) [inline src/fuzz_target.cc]: Don't use C-style casting. Instead, enforce `static_cast` or `reinterpret_cast` in C++ manner.

Comment (jserv) [reply src/fuzz_target.cc]: Run `clang-format -i` on C++ files as well.

File: .ci/fuzz.sh [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    -g -O1 \
+    -fsanitize=fuzzer,address,undefined \
+    -include src/common.h \
+    -D RV32_FEATURE_EXT_F=0 \
+    -D RV32_FEATURE_SDL=0 \
+    -D DEFAULT_STACK_ADDR=0xFFFFE000 \
+    -D DEFAULT_ARGS_ADDR=0xFFFFF000 \
+    -D FUZZER \
+    -o build/rv32emu_fuzz \
+       src/fuzz_target.cc \
```
Comment (jserv) [inline .ci/fuzz.sh]: Rename from `fuzz_target.cc` to `fuzz-target.cc` (dash instead of underscore.).

**Review by jserv — CHANGES_REQUESTED** [review]
Indent C++ source file and avoid C-style casting in C++ file.

File: docs/fuzzer.md [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+
+We are using the [LLVM Fuzzer](https://llvm.org/docs/LibFuzzer.html).
+
+The fuzzer used here is without structured input generation. Instead, we rely
+on the fuzzer to mutate the input.
+
+The initial seeds are all the ELF files in the `build` directory.
+
+## Execution
+
```
Comment (jserv) [inline docs/fuzzer.md:11]: Explain what the script does.

File: src/main.c [hunk 23]
```diff
@@ -116,6 +136,38 @@ static bool parse_args(int argc, char **args)
     int opt;
     int emu_argc = 0;
 
+#ifdef FUZZER
+    // getopt() won't work with binary data as control characters will screw the
```
Comment (jserv) [inline src/main.c]: Always use C-style comments.

File: src/main.c [hunk 24]
```diff
@@ -116,6 +136,40 @@ static bool parse_args(int argc, char **args)
...
+     * string parsing
+     */
+    int idx = 1;
+    while (idx + 1 < argc) {
+        emu_argc++;
+        char opt = args[idx][1];
+        char *optarg = args[idx + 1];
+
+        switch (opt) {
+        case 's':  // binary string
```
Comment (jserv) [inline src/main.c:151]: Always use C-style comments.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `master` branch where I revert Codacy to avoid unnecessary MISRA C checks.

File: .ci/fuzz.sh (hunk 18 above)
Comment (jserv) [reply .ci/fuzz.sh:9]: Rather than creating a separate standalone fuzzer program, we can explore the option of integrating the fuzzing components as a feature activated by a command-line option. This approach aims to minimize the necessary changes for fuzzer integration. If we opt for this method, we will need to address the linkage between the C++-based LibFuzzer and the existing C codebase.

Comment (jserv) [reply .ci/fuzz.sh:9]: See `src/feature.h` for the use of macro `RV32_HAS`.

**Comment by jserv** [comment]
Rather than creating a separate standalone fuzzer program, we can explore the option of integrating the fuzzing components as a feature activated by a command-line option. This approach aims to minimize the necessary changes for fuzzer integration. If we opt for this method, we will need to address the linkage between the C++-based LibFuzzer and the existing C codebase.

See `src/feature.h` for the use of macro `RV32_HAS`.

**Comment by henrybear327** [comment]
> Rather than creating a separate standalone fuzzer program, we can explore the option of integrating the fuzzing components as a feature activated by a command-line option. This approach aims to minimize the necessary changes for fuzzer integration. If we opt for this method, we will need to address the linkage between the C++-based LibFuzzer and the existing C codebase.
> 
> See `src/feature.h` for the use of macro `RV32_HAS`.

I am not aware of how we can merge both `main()` functions. 

My current understanding is that for the LibFuzzer to work, we will need to provide an implementation of `LLVMFuzzerTestOneInput`, and during complication, a separate `main()` will be generated and be used for driving the actions for the fuzzer.

If there is more information or hints for merging the fuzzer into the current binary, I am happy to learn! :)

**Comment by jserv** [comment]
> My current understanding is that for the LibFuzzer to work, we will need to provide an implementation of `LLVMFuzzerTestOneInput`, and during complication, a separate `main()` will be generated and be used for driving the actions for the fuzzer.

You don't have to replace the existing `main` function. Instead, check the option for running riscv-arch-tests. You can provide an option which triggers LLVM Fuzzer.

File: src/main.c (hunk 24 above)
Comment (jserv) [inline src/main.c]: Always use C-style comments.

File: src/main.c [hunk 25]
```diff
@@ -117,6 +137,40 @@ static bool parse_args(int argc, char **args)
...
+            opt_elf_string = true;
+            elf_string = (uint8_t *) optarg;
+            emu_argc++;
+            break;
+        case 'l':  // binary string len
```
Comment (jserv) [inline src/main.c]: Always use C-style comments.

File: src/main.c [hunk 26]
```diff
@@ -117,6 +137,40 @@ static bool parse_args(int argc, char **args)
...
+            opt_elf_strlen = true;
+            elf_strlen = atoi(optarg);
+            emu_argc++;
+            break;
+        case 'k':  // max execution cycle
```
Comment (jserv) [inline src/main.c]: Always use C-style comments.

**Comment by jserv** [comment]
It seems that C-only source files are compatible with libFuzzer. e.g., [libFuzzer-examples](https://github.com/eddyspaghette/libFuzzer-examples).

File: src/fuzz-target.cc [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+
+    char *arg3 = (char *) malloc(3);
+    strncpy(arg3, "-l", 3);
+    args[3] = arg3;
+    char *len_str =
+        (char *) malloc(20 + 1); /* LLONG_MIN in base 10 has 20 chars */
+    sprintf(len_str, "%zu", len);
+    args[4] = len_str;
+
+    char *arg5 = (char *) malloc(3);
```
Comment (jserv) [inline src/fuzz-target.cc:44]: Avoid C-style casting in C++ code.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `master` branch.

**Comment by jserv** [comment]
I am closing this pull request because we have not made a clear decision on how we can effectively utilize libFuzzer to identify potential issues.
